arangod/RocksDBEngine/RocksDBVPackIndex.cpp

```diff
@@ -2858,4 +2859,155 @@ void RocksDBVPackIndex::warmupInternal(transaction::Methods* trx) {
   LOG_TOPIC("499c4", DEBUG, Logger::ENGINES) << "loaded n: " << n;
 }
 
+namespace {
+
+struct RocksDBVPackStreamOptions {
+  std::size_t keyPrefixSize;
+};
+
+template<bool isUnique>
+struct RocksDBVPackStreamIterator : AqlIndexStreamIterator {
+  RocksDBVPackIndex const* _index;
+  std::unique_ptr<rocksdb::Iterator> _iterator;
+
+  RocksDBVPackStreamOptions _options;
+  VPackBuilder _builder;
+  VPackString _cache;
+  RocksDBKeyBounds _bounds;
+  rocksdb::Slice _end;
+
+  RocksDBVPackStreamIterator(RocksDBVPackIndex const* index,
+                             transaction::Methods* trx,
+                             RocksDBVPackStreamOptions options)
+      : _index(index),
+        _options(std::move(options)),
+        _bounds(
+            isUnique
+                ? RocksDBKeyBounds::UniqueVPackIndex(index->objectId(), false)
+                : RocksDBKeyBounds::VPackIndex(index->objectId(), false)) {
+    _end = _bounds.end();
+    RocksDBTransactionMethods* mthds =
+        RocksDBTransactionState::toMethods(trx, index->collection().id());
+    _iterator = mthds->NewIterator(index->columnFamily(), [&](auto& opts) {
+      TRI_ASSERT(opts.prefix_same_as_start);
+      opts.iterate_upper_bound = &_end;
+    });
+    _iterator->Seek(_bounds.start());
+  }
+
+  bool position(std::span<VPackSlice> span) const override {
+    if (!_iterator->Valid()) {
+      return false;
+    }
+
+    TRI_ASSERT(RocksDBKey::objectId(_iterator->key()) == _index->objectId())
+        << RocksDBKey::objectId(_iterator->key())
+        << " actual = " << _index->objectId();
+
+    auto keySlice = RocksDBKey::indexedVPack(_iterator->key());
+    storeKey(span, keySlice);
+    return true;
+  }
+
+  void loadKey(std::span<VPackSlice> span) {
+    _builder.clear();
+    _builder.openArray();
+    for (auto k : span) {
+      _builder.add(k);
+    }
+    _builder.close();
+  }
+
+  void storeKey(std::span<VPackSlice> into, VPackSlice keySlice) const {
+    std::size_t idx = 0;
+    for (auto k : VPackArrayIterator(keySlice)) {
+      if (idx >= _options.keyPrefixSize) {
+        break;
+      }
+      into[idx] = k;
+    }
+  }
+
+  bool seek(std::span<VPackSlice> span) override {
+    loadKey(span);
+    RocksDBKey rkey;
+    rkey.constructUniqueVPackIndexValue(_index->objectId(), _builder.slice());
```

Comment:
Potential performance optimization: we could make the RocksDBKey value an instance member, so it can be recycled to create multiple keys.